A semi-working push of our Threading feature

milljm opened the pull request:
Just committing my stuff. The lights turned off. So time to go home...

Refs #9280

Commit:
A semi-working push of our Threading feature

Refs #9280

moosebuild commented:
Job `Moose Documentation:linux-clang` on da8fad1 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9335/site

milljm commented on python/TestHarness/util.py:
```diff
@@ -562,9 +622,138 @@ def didFail(self):
+            for concurrent_testers in concurrent_tester_sets:
+                output_files_in_dir = set()
+                for tester in concurrent_testers:
+                    if self.prereqsExists() and name_to_object[tester].getTestName() not in self._skipped_tests():
```
hmmm only do this if there are prereqs? no no.. very wrong.

milljm commented on python/TestHarness/util.py:
```diff
@@ -562,9 +622,138 @@ def didFail(self):
+    # return a set of finished non-passing or will be skipped tests
+    def _skipped_tests(self):
+        skipped_failed = set([])
+        for tmp_tester in self.__testers:
```
self.__testers does not include the tester that entered this class... Then again, this class does not even get instanced if _this_ tester is skipped.

Hmmm.

moosebuild commented:
Job `App tests:linux-gnu` on da8fad1 : canceled by @milljm

haha

Commit:
Address issues

If the application is not available, we need to exit the main thread, not
the individual threads.

Fix some silliness. Typos.

Refs #9280

moosebuild commented:
Job `Moose Documentation:linux-clang` on 6882d67 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9335/site

Commit:
Handle missing applications

Allow RunApp to handle the missing application (set tester status to fail), and
allow the TestHarness to continue to run other tests that might succeed (like
python tests). The TestHarness will still bomb out once it hits MAX Failures.

Refs #9280

moosebuild commented:
Job `Moose Documentation:linux-clang` on bbd76cd wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9335/site

Commit:
Lock thread pool if tester has dependents

If the test in question has a dependent, we need to lock the thread pool while
we process this test's results. The dependent test can launch _as soon as_ this
test sets a passing status.

We should look into why we need to thread lock in this case, because according
to the tester, the last thing we set is a status. So the results file in question
should have remain unmolested until we set a passing status. But that does not
appear to be the case.

For now, if a test has dependents, lock the thread until the process has completed.

Refs #9280

moosebuild commented:
Job `Moose Documentation:linux-clang` on 172155a wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9335/site

Commit:
Handle RunApp _only_ tests differently

When a test is strictly a RunApp test (like a python test) allow RunApp to set a
success status. Continue to set a failed statuses on failure events.

This is neccessary due to our Exodiff testers (and CSVDIFF and possibly others...)
having a success status being set when in fact, the test was not entirely finished
(see processResults in Exodiff --> FileTester --> RunApp). When that happens,
sometimes a prereq thread that was waiting for this test to _PASS_, launches. Thus
calling tester.prepare(). Which _deletes_ the output file to start anew, causing
the original Exodiff test to fail with "file not found errors" just as it started
to run an exodiff command. Fun find.

Refs #9280

moosebuild commented:
Job `Moose Documentation:linux-clang` on a08eaf6 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9335/site

Commit:
Timing arguments fix

Timing arguments was casuing skipped tests to fail (operating on a
None type timing variable).

Fix all instances where timing was not taking place before this PR
as well.

Refs #9280

moosebuild commented:
Job `Moose Documentation:linux-clang` on 3d8b753 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9335/site

moosebuild commented:
Job `Test intel:linux-intel` on 3d8b753 : canceled by @milljm

ugh

moosebuild commented:
Job `External App tests:linux-gnu-external-apps` on 3d8b753 : canceled by @milljm

moosebuild commented:
Job `TBB test:linux-gnu` on 3d8b753 : canceled by @milljm

moosebuild commented:
Job `Pthreads test:linux-gnu` on 3d8b753 : canceled by @milljm

permcody commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,130 @@
+        if tester.checkRunnableBase(options):
+
+            # Ask if this test is allowed to run based on other test's circumstances
+            status_check = StatusDependency(tester, testers, options)
```
I don't think I like this name. Also, it doesn't follow style guidelines.

permcody replied:
Oh this is some sort of new object? Hmm...

milljm replied:
I suppose it doesn't have to be a new status dependency object. I could shimmy it into the TestStatus class perhaps. Or just a couple of methods not needing a class. I am surely not using it correctly by having the need to initialize it with a constructor. Overkill. Its only instanced once and thrown away...

permcody commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,130 @@
+                    tester.start_time = clock()
+                    tester.exit_code = 0
+                    tester.end_time = clock()
+                    tester.std_out = ''
```
Nope! You are digging into the Tester object from the Scheduler object to change it's internal state. The Tester should change it's own state.

permcody commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,130 @@
+        else:
+            tester.end_time = clock()
+
+    def getDependents(self, tester, testers):
```
I don't think this method has the right name. It's called `getDependents` but it returns a Boolean.

milljm replied:
hah yeah... well I removed this anyway.... turns out we don't need such a function

permcody commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,130 @@
+    # Modify the output the way we want it. Run processResults
+    def testOutput(self, tester, testers, results, options):
+        # create verbose header
+        output = 'Working Directory: ' + tester.getTestDir() + '\nRunning command: ' + tester.getCommand(options) + '\n'
```
Design: Who do you think should be in charge of formatting? The Tester, Scheduler, or TestHarness? If you ask me, I'd say the TestHarness should handle all I/O.

permcody commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,130 @@
+        # dry run? cool, just return
+        if options.dry_run or not tester.shouldExecute():
+            tester.std_out = output
+            tester.setStatus(tester.getSuccessMessage(), tester.bucket_success)
```
My encapsulation violation.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,263 @@
+        return params
+
+    ## Return this return code if the process must be killed because of timeout
+    TIMEOUT = -999999
```
This should probably be cleaned up, finally.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,263 @@
+        params['max_processes'] = self.options.jobs
+        params['average_load'] = self.options.load
+
+        # For backwards compatibitliy the RunParallel class can be initialized
```
backwards compatibility? I don't think we want to drop this design moving forward do we?

milljm replied:
I have not yet implemented a way to refrain from launching a job if x amount of CPUs are already busy.

As I recall, we talked about this, and we were not sure how to manage this situation.
-j 32 = 32 _threads_ capable of running 32 jobs. _however_ some of those jobs my ask for more than one processor...

I think we said we would deal with this at a later time once threads was actually working.

permcody replied:
Right! Let's just get this working first. 

After that we can make sure we don't have that problem by adding the total slots a job requires to a shared "slots" variable in the scheduler base. You'll have to lock that variable each time you want to inspect and write to it.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,263 @@
+            # Handle timeouts
+            self.do_TimeoutTests(tester)
+
+    def do_LongRunningTests(self, tester):
```
Doesn't follow style guidelines.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,263 @@
+
+        self.runner_pool.close()
+        self.runner_pool.join()
+        self.status_pool.close()
```
Probably should have a comment here. I think we discussed that this had to be in this order.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,263 @@
+            self._runnerGo(tester.getTestName())
+
+            # We should spin here for just a bit; A runner and status thread will rapidly hand off a
+            # job to each other for tests that can't run yet due to a prereq not being ready.
```
> each other

Really? I'm sure this is necessary but I don't think this comment explains the reason very well.

milljm replied:
I could write a TL;DR comment? =D
Hows this instead:

Because we are working with threads, there exists the possibility of having two jobs and the availability of two threads. Job A launches and takes an hour to finish. Job B depends on Job A, so it spins and checks every <sleep time> to see if it can go.

The only way that works, is if the runner pool actually tries to run (runner_pool is responsible for performing checkRunnable, and StatusDependency check thing which I agree, the name can probably be better). No matter the outcome, the job is then handed off to the status queue for finalized processing. But in this case, there is nothing to process (its still pending) so throw it back in the runner_queue... repeat. If we don't wait for just a tiny bit, that spins the processor at 100%.

Maybe the runner_queue can put things back into itself instead and assign a thread to it? (a child thread assigning more child threads) I donno... can that work?

permcody replied:
I think sleep is just fine. I just couldn't remember the usage, now I recall.

permcody reviewed (commented):
Just a partial review for now. This is coming along but there is still a few design issues and naming convention problems. Still, this is a big project so keep at it.

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -11,7 +11,9 @@
-from RunParallel import RunParallel
+from Scheduler import Scheduler
+from CSVDiffer import CSVDiffer
+from XMLDiffer import XMLDiffer
```
Those were unused and I removed them in #9330

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -145,78 +150,73 @@ def findAndRunTests(self, find_only=False):
                 self.writeFailedTest.close()
-            print('\nExiting due to keyboard interrupt...')
-            sys.exit(0)
+            print '\nExiting due to keyboard interrupt...'
```
Yes, we really should make sure that new code uses newer print statements.

milljm replied:
Sorry, I got overzealous trying to solve conflicts

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -444,240 +339,113 @@ def checkExpectError(self, output, expect_error):
+        # The test has not yet finished but there is a status change we want to display
         else:
-            return False
+            print printResult(tester, tester.getStatusMessage(), clock() - tester.getStartTime(), self.options)
```
`printResult` doesn't actually print anything. This should also use new `print()`

milljm replied:
the printResult function formats the output to fit the screen, color the text, etc etc and returns that formatted output to be 'printed' (print _formatted_text_)

milljm replied:
printResults should probably be named formatResults or something...

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -127,10 +129,13 @@ def __init__(self, argv, app_name, moose_dir):
+        checks['method'] = method
+
+        # This is so we can easily pass checks around to any scheduler plugin
+        self.options._checks = checks
```
This was to clean up the requirements as we pass the tester around. When ever we perform a checkRunnable, we needed to pass checks, failed_tests, and options.

Now we just pass options.

milljm replied:
No? Better way? Because I don't know how to get all this info into the Tester with out passing it around. We could certainly go back to passing three arguments. I don't really care.

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -145,78 +150,73 @@ def findAndRunTests(self, find_only=False):
+                # walk into directories that aren't contrib directories
+                if "contrib" not in os.path.relpath(dirpath, os.getcwd()):
+                    for file in filenames:
+                        # set cluster_handle to be None initially (happens for each test)
```
Old PBS remnants... I'll remove it

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -444,240 +339,113 @@ def checkExpectError(self, output, expect_error):
+    def handleTestStatus(self, tester):
+        # The test has finished regardless of status
+        if tester.isFinished():
+            if tester.isSilent() or (tester.isDeleted() and not self.options.extra_info):
```
Somewhere in the code, we're going to need to know when and when not to print. The scheduler does not decide this. It just throws a job in the status queue, which in turns calls back to the TestHarness.

Maybe we can always store the finalized output only in the tester... and _always_ print that object no matter what. That way the tester controls when/how/if stuff gets printed.
```python
def handleTestStatus(self, tester):
    footer = keep_track_of_totals(tester)
    if tester.getOutput() is not None:
        print tester.getOutput()

def printFinalResults(self):
    print footer
```

milljm commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,130 @@
+    @staticmethod
+    def validParams():
+        params = Scheduler.validParams()
+        params.addRequiredParam('scheduler',    'RunParallel', "The name of this scheduler")
```
I am going to need your help on this one. I still don't understand the MOOSE factory warehouse parser thingamajig

permcody replied:
Just get rid of it. You shouldn't need it at all. The inheritance from MooseObject should already do what you need to do. You can build any object without this parameter. I'd have to look at where you actually use this parameter.

milljm replied:
I need it. But if you are sure, I need your help to solve the error I get when I remove it:
```python
Traceback (most recent call last):
  File "./run_tests", line 26, in <module>
    TestHarness.buildAndRun(sys.argv, app_name, MOOSE_DIR)
  File "/ssd/milljm/moose/python/TestHarness/TestHarness.py", line 30, in buildAndRun
    harness = TestHarness(argv, app_name, moose_dir)
  File "/ssd/milljm/moose/python/TestHarness/TestHarness.py", line 137, in __init__
    self.initialize(argv, app_name)
  File "/ssd/milljm/moose/python/TestHarness/TestHarness.py", line 498, in initialize
    self.scheduler = self.factory.create(plugin_params['scheduler'], self, plugin_params)
  File "/ssd/milljm/moose/python/FactorySystem/InputParameters.py", line 49, in __getitem__
    return self.valid[key]
KeyError: 'scheduler'
```

permcody replied:
Look at the stack trace (Line 498 in TestHarness.py): You are still asking for this key explicitly, but it's your code, not the Factory, Parser, or Warehouse.

milljm replied:
thats all voodoo to me, but I think I got it.
```python
        self.scheduler = self.factory.create(scheduler_plugin, self, plugin_params)
```

Commit:
Do not oversubscribe the machine

Keep track of workers when tests launch to prevent the machine from become saturated
with jobs.

Still need to implement the possibility of a 1 core (-j 1) event.

Refs #9280

Commit:
Change printResult to formatResult

The name was confusing, as it didn't print anything.

Refs #9280

moosebuild commented:
Job `Moose Documentation:linux-clang` on 5d54519 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9335/site

Commit:
Implement a DAG dependency scheduler

Use a DAG to control when tests are allowed to run.

Refs #9280

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -151,23 +157,142 @@ def statusGo(self, tester_name):
+
+        # Add our local DAG to the global shared dag
+        #with self.thread_lock:
+        self.shared_testers_dag[testers[-1].getTestDir()] = tester_dag
```
I was fooling around with not using a lock. We probably should for this case. I've been reading its not necessary when creating a new node?

permcody replied:
Where did you read that? I'm not so sure this is true unless you can point me to some documentation.

permcody replied:
Ah, I see, you aren't talking about adding a node to the dag. You are talking about adding a node to the dict. I don't like how you are getting the tester directory here. Why the last one, why not the first one? To somebody who is less familiar with the code this is confusing.

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -151,23 +157,142 @@ def statusGo(self, tester_name):
+
+        # Set a pending status on these jobs _before_ launching them
+        for job in ind_jobs:
+            self.scheduled_name_to_object[job].setStatus('QUEUED', self.scheduled_name_to_object[job].bucket_pending)
```
Lock or not? Not sure if we need it.

permcody replied:
Can you point me to the doco? I'll help you figure it out.

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -227,74 +352,120 @@ def doTimeoutTests(self, tester):
+        # Test is Finished (failed, passed, skipped, silent, etc)
         if tester.isFinished():
-            # And the running caveat and
+            # And the running caveat
```
lol. I donno. I'll someday fix this typo correctly

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -227,74 +352,120 @@ def doTimeoutTests(self, tester):
+
+                # Get the next list of runnable nodes (jobs)
+                tester_dag.delete_node(tester_name)
+                next_jobs = tester_dag.ind_nodes()
```
should leave a cooler comment here.

permcody replied:
So you don't need to get the `ind_nodes()` inside the lock. I had directed you to do this last week and if it had worked properly you wouldn't have to do the check to see if the job was running before adding it to the queue. However as we discussed, you will get the runnable jobs and running jobs from the dag every time you ask since we don't remove the nodes until the test is finished. Now that we know that, you can just remove the lock.

You only need to lock the inserts and deletes.

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -227,74 +352,120 @@ def doTimeoutTests(self, tester):
-            # job to each other for tests that can't run yet due to a prereq not being ready.
-            sleep(0.01)
+            # job to each other for tests that can't run yet due to not enough slots being available
+            #sleep(0.01)
```
I was testing a no sleep. @permcody Its as you thought... I don't think we need it. I had a print statement here that never really fired. Which is awesome! That means hardly any testers were pending.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 47aeaec wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9335/site

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+                self.scheduled_name_to_object[tester.getTestName()] = tester
+
+            # Create a node for this tester if not already exists (prereq iteration creates nodes)
+            if not tester_dag.node_exists(tester.getTestName()):
```
Another idea for you. You might add another method to the DAG that just adds nodes all the time so you don't have to check first. If you find yourself always doing this, then it would be useful to have that utility.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+                    skipped_or_failed_tests.add(tester)
+                    continue
+
+                if not tester_dag.node_exists(prereq):
```
Here you are asking before creating again.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+                try:
+                    tester_dag.add_edge(prereq, tester.getTestName())
+
+                # Handle Cyclic errors (tester status has already be applied)
```
Where do we find cyclic tests? Is this in the checkRaceCondition stuff? Maybe we should just combine that into building the DAG.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+
+                # Handle Cyclic errors (tester status has already be applied)
+                except dag.DAGValidationError:
+                    pass
```
This feels wrong. It sounds like you never expect to hit this error so why would you pass?

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+                skipped_or_failed_tests.add(tester)
+
+        # Discover skipped tests and remove downstream tests that require this skipped test
+        for tester in skipped_or_failed_tests:
```
This whole for-loop could go away once you fixup the deleteDownstream method.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+        ind_jobs = tester_dag.ind_nodes()
+
+        # Maybe all jobs in this group got skipped?
+        if ind_jobs == []:
```
Careful here. While this effectively works it's logically incorrect. You know that because there are no independent nodes, there aren't any nodes. The only other case would be a cyclic graph which you've already handled. Logically it would be better to short circuit if the graph is empty instead.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+        self.status_pool.join()
+
+    # Method to do things needed to actively running tests.
+    # NOTE: We are in a thread locked state, so be quick
```
Thanks for the comment here, but I'd avoid locking around a whole routine. It's actually two or three routines down before you actually do anything with the data structure. You should only need to lock the actual changes to the structure.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+            if clock() - tester.getStartTime() > min_time:
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('RUNNING...', tester.bucket_pending)
+                self.harness.handleTestStatus(tester)
```
Yuck - You just locked this whole call back to a different object that doesn't even know about threading...

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+            if clock() - tester.getStartTime() > float(tester.getMaxTime()):
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('TIMEOUT', tester.bucket_fail)
+                self.harness.handleTestStatus(tester)
```
Lock here with a callback to an object that doesn't even know about threading. Keep your locks very local, right where you need them.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+    # TODO: Possibly look into semaphore functionality for slot allocation
+    def checkAvailableSlots(self, tester):
+        # See if we can fit this job into our busy schedule
+        if self.workers_in_use + tester.getProcs(self.options) <= self.job_slots:
```
`self.workers_in_use` isn't thread safe. I can see that you are locking the call to this routine but again that's bad style. Most of what's in here can be moved to the schedule tab or be done somewhere before we get to the threading region. For instance the oversized calculation doesn't need to happen here. Try to minimize the work and variable you need to access when working with threads.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+        # Check for insufficient slots -hard limit (skip this job)
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and not self.soft_limit:
+            tester.setStatus('insufficient slots', tester.bucket_skip)
```
Both of these checks can happen before you get to the threaded region. Then you can pull the dynamic check out and just put the logic directly in where you are using it right under the thread lock.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+
+            # Remove this finished tester from the DAG and active job set, get the next node, and assign it to
+            # a runner thread
+            with self.thread_lock:
```
Don't need the lock here. You only need it where you start modifying the dag. Keep the locks as close to the operations they are protecting as possible.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+                # Handle dependencies that will fail if this test did not successfully finish but only if we care
+                # about prereqs
+                if not tester.didPass():
+                    self.deleteDownstreamTests(tester_dag, tester)
```
You'd need the lock here and just below where you actually delete.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+
+                # Loop through and determine if we need to assign a thread to this new job
+                new_tmp_jobs = []
+                for job in next_jobs:
```
This isn't as clean as I had hoped. I think you should document why you have to do this. (i.e. the DAG will give you tests that might already be in the queue because we don't remove running jobs).

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+            # The runner when performing checkRunnableBase, didn't fail or skip the test. But it couldn't
+            # run for other reasons (not enough slots, etc). So place it back in the runner queue
+            if tester_name not in self.active_jobs:
+                self.runnerGo(tester_name)
```
Not crazy about this name, doesn't tell use anything useful. Maybe `insertRunnerQueue`?

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+            if tester.isPending():
+                raise SchedulerError('Derived Scheduler can not return a pending status!')
+
+            # Making modifications to the tester requires a thread lock
```
Why???!?!?!?!?!?!?!?!?!?!?

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+                    tester.setEndTime(clock())
+
+                # Recover our available slots
+                self.workers_in_use = max(0, self.workers_in_use - tester.getProcs(self.options))
```
why max here? It better always be zero unless you are doing something screw for oversized jobs?

permcody commented on python/TestHarness/testers/CheckFiles.py:
```diff
@@ -23,9 +23,13 @@ def getOutputFiles(self):
+        reason = FileTester.processResults(self, moose_dir, retcode, options, output)
 
         specs = self.specs
+        if options.enable_recover and specs['skip_checks']:
```
Ugh... I don't like this recover crap bleeding into the individual Testers. Can we handle this at the base class?

milljm replied:
haha I know. It made me mad when I ran into this.

permcody commented on python/TestHarness/testers/Tester.py:
```diff
@@ -61,21 +62,28 @@ def __init__(self, name, params):
+        self.__start_time = clock()
+        self.__end_time = None
+        self.__exit_code = 0
+        self.__std_out = ''
```
This is good! You are still setting these from outside the Tester though. I commented on that earlier.

permcody commented on python/TestHarness/util.py:
```diff
@@ -96,17 +98,36 @@ def runCommand(cmd, cwd=None):
     return output
 
+# Execute a command and return the process, temporary output_file and time the process was launched
+def returnCommand(tester, command):
```
Weird name: returnCommand()? Why not launchProcess(). This is really close to "getCommand()"

permcody reviewed (changes requested):
This is getting close despite the 44 comments I just made. I wanted to be thorough and I don't see any showstoppers. Just lots of little cleanup.

Commit:
Just another Threading commit

Add a tester data container object class to contain the tester, which
provides a way to separate things we believe should be controlled by
the Scheduler class instead (file output handlers, start/end times)

Bring the printing function out of the Tester object and back into the
TestHarness.

Optimized the way threads launch jobs based on the DAG.

Refs #9280

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+
+    # Loop through the testers and add them to the runner pool for execution
+    def schedule(self, testers):
+        # We didn't recieve anything... pesky TestHarness
```
It doesn't happen. I was just being carful. :)

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+        self.clearAndInitializeJobs()
+
+    ## Clear and Initialize the scheduler queue
+    def clearAndInitializeJobs(self):
```
True... and yes. Its absolutely going to be needing when we introduce the QueueManager. But maybe with this newly improved DAG thingy, I may never need to clear anything out again.

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+from MooseObject import MooseObject
+from Queue import Queue
+from signal import SIGTERM
+from util import *
```
agreed. Its caught me several times already.

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -873,15 +522,14 @@ def parseCLArgs(self, argv):
         if self.options.failed_tests:
             with open(failed_tests_file, 'r') as tmp_failed_tests:
-                self.test_list = tmp_failed_tests.read().split('\n')
+                self.options._test_list = tmp_failed_tests.read().split('\n')
```
The Tester is using this during checkRunnableBase. It works along the same means as --re

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98077/) on 2132df1 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9335/site)

Commit:
Address comments

Add a load_average routine to the scheduler.

Remove options.enable_recover from all the testers and have
RunParallel handle it at a base level.

Refs #9280

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,471 @@
+
+        # Status queue (we put jobs we want to display a status for in this)
+        # status_queue.put( (tester, clock(), subprocess_object) )
+        self.status_queue = Queue()
```
Not sure why this comment is labeled as addressed. But I removed the "clear stuff" method all together.

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98104/) on 8eda17e wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9335/site)

Commit:
Address more comments

Create a new file to hold the TestData class. <-- This is a WIP.

Refs #9280

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98340/) on 76f4a60 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9335/site)

permcody commented:
@milljm - you  don't need to put the component label on your PRs. They are on the tickets to which PRs are linked. I'm trying to keep the labels on PRs down a bit.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,491 @@
+            self.deleteDownstreamTests(skipped_tester_data)
+
+            # Delete this node
+            tester_dag.delete_node_if_exists(tmp_tester.getTestName())
```
Your new method which you will create `delete_sub_tree()` should probably delete the whole subtree rooted at the given node. If you do that, you won't have to delete the downstream, then the current node. It'll all be one operation.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,491 @@
+
+        # With the job properly status'd above... launch them all asynchronously
+        for job in ind_jobs:
+            self.assignRunner(job)
```
I think your assignRunner() routine should take a list of items (something iterable). Inside that method you'll perform the loop. Still don't like the methodName "assignRunner" the reason is that you really aren't assigning a runner. You don't really have the authority to do that, that's technically the ThreadPool's job. Instead you are queuing jobs to run. 

How about you make a routine that takes both the "list" and the "ThreadPool" object or a name and you call the method `queueJobs()`

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,491 @@
+        # Check if we have available slots
+        if self.checkAvailableSlots(tester):
+            can_run = True
+            with self.thread_lock:
```
NOPE! You created a race condition here that's hard to see because of the subroutine you called. You have to be very careful with locks and subroutines. In `checkAvailableSlots()`, you lock and inspect at the shared resource `workers_in_use` to figure out whether you can run. However, you then let go of the lock as you exit the routine after you've answered the question. Another thread can sneak in there and also think it's OK to start. You have to keep the inspection and writing of the resource together. This is paramount.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,491 @@
+                self.active_jobs.add(job)
+
+                # Record the amount of slots we're about to consume
+                self.workers_in_use += tester.getProcs(self.options)
```
Race condition for this variable. Instead, make your method names explicit if they must be atomic: `checkAndReserveSlots()`

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,491 @@
+
+        # We can not run just yet (slots full)
+        else:
+            sleep(0.03)
```
interesting sleep time choice: 3/100 of a second might be a little short. On the other hand, users really shouldn't be oversubscribing their machine. Not worried about this yet as I still think there is a way to avoid this. I'll talk to you about it later...

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,491 @@
+
+        # Launch these next set of jobs
+        for next_job in next_job_group:
+            self.assignRunner(next_job)
```
Yeah - this is all closer then sending in the DAG. The items in the queue should be Testers (or more accurately TesterDatas) of course in Python under the hood that means references to the actual items which sit in the DAG.

Commit:
Cleaned up the scheduler routine a bit

Addressed comments.

Refactored several methods that were each designed to find certain
errors (like race conditions and unknown dependencies) into one
method which uses a DAG to perform all of it. By doing so, I added
more detailed exception handling in the DAG class.

Refs #9280

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+        self.status_pool = ThreadPool(processes=1)
+
+        # Thread Locking
+        self.thread_lock = threading.RLock()
```
I did some reading on this. "R" is for reentrant which means the thread that locks it can continue to pass through barriers protected with the same lock. My opinion is that unless this is something we need, we probably shouldn't use it because it might actually obscure logic problems with locking. Use "Lock"instead.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+        self.name_to_dag = {}
+
+        # A simple queue increment to keep track of jobs
+        self.job_queue_count = 0
```
I see `workers_in_use` and `job_queue_count`. I don't have a clear picture of what these two different counts are for from the description. You may want to beef up those descriptions.

permcody commented on python/contrib/dag/__init__.py:
```diff
@@ -0,0 +1,224 @@
+            i += 1
+        return filter(lambda node: node in nodes_seen, self.topological_sort(graph=graph))
+
+    def delete_downstreams(self, node, graph=None):
```
👍

permcody replied:
You might want to put comments on the methods that you added so we know where the contrib code ends and yours begins.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+                or (self.options.dry_run and not tester.isSilent()):
+
+                # Ask the DAG to delete and return the downstream jobs associated with this job
+                failed_job_containers.update(job_dag.delete_downstreams(job_container))
```
👍

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+        return skipped_failed
+
+    # Delete and return downstream nodes based on tester status and TestHarness options
+    # TODO: change the method name, as it doesn't always delete downstream tests
```
Well, I'd suggest that you go ahead and rename it then: Maybe `processDownstreamTests`?

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+
+    # Iterate over supplied tester data container dict and add edge nodes to the DAG (create dependencies)
+    # Note: This method ignores all errors
+    def createEdgeNodes(self, job_container_dict):
```
I don't like this name. It's not really descriptive in terms of a scheduler (This method name is more appropriate for a DAG object). You are creating job dependencies. Go ahead and call it that.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+
+                # we do not care about missing prereqs
+                if prereq not in job_container_dict.keys():
+                    continue
```
This logic is a little odd (the negative test and immediate continue). Try this instead and get rid of the else clause:

```
if prereq in joob_container_dict.keys():
   job_dag.add_node_if_not_exists(job_container_dict[prereq])
   ...
```

milljm replied:
This is would cause a KeyError, because the prereq value listed in the 'tests' file is invalid, and was already detected by the checkGroupFailures method ('unknown dependency'). So due to 'prereq' being a bogus value at the moment, I continue to the next iteration.

permcody replied:
No, the logic I'm showing is exactly equivalent and cleaner. Now if you were using the prereq variable elsewhere (outside of the if statement), it'd be a different story but you are not. This is a very small loop with only one branch statement.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+            tester = job_container.getTester()
+
+            # If this tester is not runnable, continue to the next tester
+            if not tester.getRunnable(self.options):
```
Another "if not foo then continue else do stuff". You can always switch it around and just do "if foo then do stuff"

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+
+                # Skipped/Silent/Deleted Testers fall into this catagory, caused by 'job_container' being skipped during the first iteration above
+                except dag.DAGEdgeDepError:
+                    pass
```
I might structure this differently, put the pass case at the bottom. That way you work your way through all the cases that you are going to actually handle followed by the cases that are left. This is fine though if you like it the way it is.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+                    tester.setStatus('unknown dependency', tester.bucket_fail)
+                    failed_or_skipped_testers.add(tester)
+
+        # With a working DAG created above (even a partial one), discover race conditions with remaining runnable testers.
```
👍

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+
+        # Iterate over the items in our non_runnable_jobs and handle any skipped dependencies
+        # Note the copy method, as this may add more skipped tests to the original set as we learn about failed downstream jobs
+        for job in non_runnable_jobs.copy():
```
Clever, maybe too clever... I've been down this road before and know why this is here.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+                tester.setStatus('QUEUED', tester.bucket_pending)
+
+        # Queue runnable jobs
+        self.queueJobs(tester_job_containers=job_list)
```
This is looking much better!

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+        self.status_pool.join()
+
+    # Create thread timers for various timing tasks (timeouts, long running)
+    # TODO:  to clever? separate these? The idea was we can add as many different timers as we want
```
This is OK - You can remove the TODO method.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+            sleep(1.0)
+
+    # Return bool if we have enough resouces to run the job
+    def checkAvailableSlots(self, job_container):
```
`reserveSlots()` is probably a better name. This method will block and more importantly it actually reserves the slots so "check" isn't the right name.

permcody replied:
Hmm - something seems wrong with this method. What about the case where there are enough total slots but not enough open slots? None of those cases are covered which means you return false. In the place where you call this method you end up in never will run case. I suspect you aren't hitting this case often because you are generally running one slot per job (most of the time). If you ran a parallel run_tests though you'd be running out of slots while there were lots of extra workers. Maybe I'm missing something but this seems just wrong.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+        # Make sure we are complying with the requested load average
+        self.satisfyLoad()
+
+        with self.thread_lock:
```
Much better!

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+            self.queueJobs(tester_job_containers=next_job_group)
+
+        # Not enough slots to run the job
+        else:
```
Yeah - this can't be right. One check on the currently running slots and you end up in too big land even if it would fit later.

permcody reviewed (changes requested):
Getting much closer! Only one big logic problem I see. It's about time that we start working with this on our systems and really running it through some stress tests.

Commit:
Address Comments

Refs #9280

brianmoose commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,93 @@
+
+## This class provides an interface to run tester commands and do something with the output
+#
+# To use this class, call the .run() method with the tester_data container
```
docstring

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -393,337 +325,127 @@ def appendRecoverableTests(self, testers):
-            self.handleTestResult(tester, '', result, 0, 0, True)
-            test_completed = True
+    # handleTestStatus calls the tester's ability to print its current status to the screen, and handles
+    # the finalized status for tallying up the totals based on those statuses (failed, success, etc)
```
docstring

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -393,337 +325,127 @@ def appendRecoverableTests(self, testers):
-            lines = output.split('\n');
+                if self.options.file:
+                    self.file.write(util.formatResult( tester, result, timing, self.options, color=False) + '\n')
+                    self.file.write(output)
```
output not defined

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -393,337 +325,127 @@ def appendRecoverableTests(self, testers):
+                    fname = os.path.join(tester.getTestDir(), tester.getTestName().split('/')[-1] + '.' + result[:6] + '.txt')
+                    f = open(fname, 'w')
+                    f.write(util.formatResult( tester, result, timing, self.options, color=False) + '\n')
+                    f.write(output)
```
output not defined

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -749,20 +471,14 @@ def cleanup(self):
-                             colored=self.options.colored, code=self.options.code ))
-            if self.options.pbs:
-                print('\nYour PBS batch file:', self.options.pbs)
+            print util.colorText( summary % (self.num_passed, self.num_skipped, self.num_failed),  "", html = True, \
```
use print()

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -393,337 +325,127 @@ def appendRecoverableTests(self, testers):
-                print(printResult(test, result, timing, start, end, self.options))
+            print '\n\nFinal Test Results:\n' + ('-' * (util.TERM_COLS-1))
+            for (test, result, timing) in sorted(self.test_table, key=lambda x: x[2], reverse=True):
+                print util.formatResult(test, result, timing, self.options)
```
use print() here and above

brianmoose commented on python/TestHarness/TestHarness.py:
```diff
@@ -878,15 +606,14 @@ def parseCLArgs(self, argv):
         # If attempting to test only failed_tests, open the .failed_tests file and create a list object
         # otherwise, open the failed_tests file object for writing (clobber).
-        self.test_list = []
+        test_list = []
```
this doesn't seem to be used

brianmoose commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,93 @@
+from timeit import default_timer as clock
```
not used

brianmoose commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,93 @@
+from timeit import default_timer as clock
+
+from Tester import Tester
```
not used

brianmoose commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,93 @@
+
+from Tester import Tester
+from Scheduler import Scheduler
+from MooseObject import MooseObject
```
not used

brianmoose commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,93 @@
+from Scheduler import Scheduler
+from MooseObject import MooseObject
+import util
+import platform, time
```
not used

brianmoose commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,93 @@
+import util
+import platform, time
+
+import os
```
not used

brianmoose commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,93 @@
+    # The run method should be blocking until the test has completed _and_ the results have been
+    # processed (with tester.processResults()). When we return from this method, this test will
+    # be considered finished, the output file will be closed, and a worker will immediately place
+    # this job in the status queue, to have its status printed to the screen.
```
docstring

brianmoose commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -0,0 +1,93 @@
+        else:
+            tester.setStatus('ERROR LAUNCHING JOB', tester.bucket_fail)
+
+    # Modify the output the way we want it. Run processResults
```
docstring

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,526 @@
+import util
+from multiprocessing.pool import ThreadPool
+import threading # for thread locking
+import multiprocessing # for timeouts
```
multiprocessing, deque, SIGTERM, sys, re, copy, util are not used

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,526 @@
+    def validParams():
+        params = MooseObject.validParams()
+        params.addRequiredParam('average_load',          64.0, "Average load to allow")
+        params.addRequiredParam('max_processes',         None, "Hard limit of maxium processes to use")
```
Why are these spaced so weird?

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,526 @@
+        for tester_data in self.tester_datas:
+            tester_data.killProcess()
+
+    ## Run the command asynchronously
```
docstring on all these functions

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,526 @@
+    # return bool if we want to ignore prereq requirements
+    def skipPrereqs(self):
+        if self.options.ignored_caveats:
+            caveat_list = [x.lower() for x in self.options.ignored_caveats.split()]
```
not used

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,526 @@
+
+                try:
+                    # This may produce a KeyError (unknown dependency)
+                    prereq_tester = job_container_dict[prereq].getTester()
```
not used

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,526 @@
+    # Remove incoming job and get next available set of jobs
+    def getNextJobGroup(self, job_container):
+        with self.thread_lock:
+            tester = job_container.getTester()
```
not used

brianmoose commented on python/TestHarness/schedulers/TesterData.py:
```diff
@@ -0,0 +1,146 @@
+import re, os, sys, platform
```
sys not used

brianmoose commented on python/TestHarness/schedulers/TesterData.py:
```diff
@@ -0,0 +1,146 @@
+import re, os, sys, platform
+from timeit import default_timer as clock
+from signal import SIGTERM
+from time import sleep
```
sleep not used

brianmoose commented on python/TestHarness/schedulers/TesterData.py:
```diff
@@ -0,0 +1,146 @@
+        self.__end_time = clock()
+        self.__std_out = ''
+
+    # Return the tester object
```
docstring here and other methods in this class

brianmoose commented on python/TestHarness/testers/CSVDiff.py:
```diff
@@ -17,9 +17,10 @@ def getOutputFiles(self):
 
     def processResults(self, moose_dir, retcode, options, output):
-        output = FileTester.processResults(self, moose_dir, retcode, options, output)
+        reason = FileTester.processResults(self, moose_dir, retcode, options, output)
```
reason not used

brianmoose commented on python/TestHarness/testers/Exodiff.py:
```diff
@@ -44,7 +44,7 @@ def processResultsCommand(self, moose_dir, options):
 
     def processResults(self, moose_dir, retcode, options, output):
-        output = FileTester.processResults(self, moose_dir, retcode, options, output)
+        reason = FileTester.processResults(self, moose_dir, retcode, options, output)
```
reason not used

brianmoose commented on python/TestHarness/testers/RunApp.py:
```diff
@@ -1,7 +1,7 @@
 from Tester import Tester
 import util
-from RunParallel import RunParallel # For TIMEOUT value
+from Scheduler import Scheduler # For TIMEOUT value
```
not used, neither is sys above

brianmoose commented on python/TestHarness/testers/RunApp.py:
```diff
@@ -180,92 +175,10 @@ def getCommand(self, options):
-
-
-    def processResults(self, moose_dir, retcode, options, output):
+    # Set a failure status for expressions found in output
```
docstring

brianmoose commented on python/TestHarness/testers/Tester.py:
```diff
@@ -2,6 +2,7 @@
 import util
 from InputParameters import InputParameters
 from MooseObject import MooseObject
+from timeit import default_timer as clock
```
clock, InputParameters not used

brianmoose commented on python/TestHarness/testers/Tester.py:
```diff
@@ -96,9 +98,30 @@ def getTestName(self):
+    def getMaxTime(self):
+        return self.specs['max_time']
+
+    # A cached method to return if the test can run
```
docstring here and other methods

brianmoose commented on python/TestHarness/util.py:
```diff
@@ -2,6 +2,8 @@
 from mooseutils import colorText
 from collections import namedtuple
+from tempfile import TemporaryFile
+from timeit import default_timer as clock
```
not used

brianmoose commented on python/TestHarness/util.py:
```diff
@@ -117,17 +119,36 @@ def runCommand(cmd, cwd=None):
+        # On Windows, there is an issue with path translation when the command is passed in
+        # as a list.
+        if platform.system() == "Windows":
+            process = subprocess.Popen(command,stdout=f,stderr=f,close_fds=False, shell=True, creationflags=CREATE_NEW_PROCESS_GROUP, cwd=tester.getTestDir())
```
I think this needs to use `subprocess.CREATE_NEW_PROCESS_GROUP`

brianmoose commented on python/TestHarness/util.py:
```diff
@@ -450,6 +470,46 @@ def deleteFilesAndFolders(test_dir, paths, delete_folders=True):
                     # TL;DR; Just pass...
                     pass
 
+# Check if test has any redirected output, and if its ready to be read
```
docstring here and elsewhere

brianmoose reviewed (changes requested):
Cosmetic changes.
Mostly caught by pyflakes

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,526 @@
+
+        for job in status_job_containers:
+            # Assign a thread to print some status
+            self.status_pool.apply_async(self.statusWoker, (job,))
```
What is a Woker?

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,526 @@
+            self.status_pool.apply_async(self.statusWoker, (job,))
+
+    # Status processing (inform TestHarness of a status)
+    def statusWoker(self, job_container):
```
Woker?

brianmoose commented on python/TestHarness/util.py:
```diff
@@ -117,17 +119,36 @@ def runCommand(cmd, cwd=None):
+        else:
+            process = subprocess.Popen(command,stdout=f,stderr=f,close_fds=False, shell=True, preexec_fn=os.setsid, cwd=tester.getTestDir())
+    except:
+        print "Error in launching a new task", command
```
print() here and elsewhere in this file

brianmoose commented on python/TestHarness/util.py:
```diff
@@ -151,17 +172,16 @@ def printResult(tester, result, timing, start, end, options, color=True):
         f_result += ' [' + '%0.3f' % float(timing) + 's]'
     if options.debug_harness:
-        f_result += ' Start: ' + '%0.3f' % start + ' End: ' + '%0.3f' % end
+        f_result += ' Start: ' + '%0.3f' % start + ' End: ' + '%0.3f' % tester.getEndTime()
```
start not defined

brianmoose commented on python/TestHarness/util.py:
```diff
@@ -450,6 +470,46 @@ def deleteFilesAndFolders(test_dir, paths, delete_folders=True):
+    return output
+
 # See http://code.activestate.com/recipes/576570-dependency-resolver/
 class DependencyResolver:
```
Do we still want to have DependencyResolver and ReverseReachability? It doesn't seem we use it anywhere...

Commit:
wip

Commit:
Address Comments

Adding thread exception handling.
Refs #9280

moosebuild commented:
Job [TBB Test](https://www.moosebuild.org/job/103002/) on c54d03c : invalidated by @milljm

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+                    try:
+                        job_dag.add_edge(job_container_dict[prereq], job_container)
+
+                    # We do not care about cyclic errors, they have been handled already
```
Hmm, I argue that you do care about them here. As you said, they have been handled already so if you were to hit one, something very bad and unexpected would have happened. I suggest that you allow this to crash because if it ever does - we'll have to investigate.

permcody replied:
Also, you are building up the same DAG here and in the method below. You should be able to build it once and then just use it after you filter out all the bad tests. e.g. In the method below where you build `temp_dag`, you should hang onto it in your dictionary. No need to build it twice!

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+        Status pool will always have the last item needing to be 'printed', we close and join
+        the Tester pool first, and then we do the same to the Status pool.
+        """
+        while self.job_queue_count > 0:
```
Can this ever get stuck? The answer might be no, but I'm just wondering if the `job_queue_count` is guaranteed to decrease over time.

milljm replied:
Its not guaranteed to decrease over time per say... But every tester that enters the Scheduler exits through the status queue no matter what its status is (silent, deleted, skipped, etc). And its the status queue's job to decrement this value when the incoming tester is labeled as 'finished'. 

There is a case where the tester may not set a status. Thats what we would call a TIMEOUT. We have a thread timer for this case. The timeout thread timer will set a finished status, kill the process, and send this test to the status queue. Which will be handled as described above.

We also have an exception handler in the event that there is a mismatch between how many testers the TestHarness is asking the Scheduler to run, verses how many the Scheduler actually did run (plus the ones that were skipped/failed). If this number does not match, we raise the exception.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+
+        # Iterate over the items in our non_runnable_jobs and handle any skipped dependencies
+        for job in non_runnable_jobs.copy():
+            additionally_skipped = self.processDownstreamTests(name_to_job_container[job.getTestName()])
```
Replace `name_to_job_container[job.getTestName()]` with `job` -> they should be equivalent.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+        for job in non_runnable_jobs.copy():
+            additionally_skipped = self.processDownstreamTests(name_to_job_container[job.getTestName()])
+            non_runnable_jobs.update(additionally_skipped)
+            job_dag.delete_node_if_exists(name_to_job_container[job.getTestName()])
```
Replace again with `job`

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+            non_runnable_jobs.add(name_to_job_container[failed_tester.getTestName()])
+
+        # Create dependencies in the DAG shared between this group of testers
+        self.createDependencies(name_to_job_container)
```
Should be able to delete this line. Might need minor update so that you actually save the dag the first time you build it.

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,539 @@
+        tester = job_container.getTester()
+
+        # comply with load average
+        self.satisfyLoad()
```
Minor: Maybe make this optional with the `-l` option.

Commit:
Address Comments

Use only one dag instead of creating two. Split up the
group failure check and race condition check.

Add a cloning method to the DAG class.

Refs #9280

Commit:
Removed Pending status from TestHarness

Because we are removing pending status (for now, but will be re-
introduced when we add back in the Queue Manager aka: PBS), this
unit test needs to change.

Refs #9280

Commit:
Handle empty list of testers

Some test files have their test entirely commented out. This
generates an empty list, which causes the Scheduler to fail.

Example:

[Test]
  # [./test]
  # [../]
[]

Refs #9280

Commit:
Addressing Comments

Refs #9280

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -136,48 +136,27 @@ def processDownstreamTests(self, job_container):
 
+        # Arbitrarily obtain a dag object from one of the job containers as they
+        # are shared between each job_container
+        arbitrary_job = job_container_dict.iteritems().next()[1]
```
What's the [1] index?

milljm replied:
A job_container (TesterData), which contains the shared DAG we need to work with

milljm replied:
getting rid of this garbage. I didn't like it either

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -0,0 +1,532 @@
+        if self.tester_pool._state:
+            return
+
+        job_dag = dag.DAG()
```
Put a comment here as to why you are creating the dag outside of the constructor of the TesterData (i.e. shared). Also, consider passing it into the soon to be created "buildDAG" routine.

permcody reviewed (commented):
Looks pretty good - will be going in very soon. Last chance @rwcarlsen, @brianmoose.

Commit:
Addressing Comments

Pass job_dag. Thats just a cleaner way to do it.
Rename the un-popular checkGroupFailers method.

Refs #9280

milljm commented:
Yeah... cause I need to start on the QueueManager side of this soon (PBS)!

permcody commented:
Looks like we aren't getting output for some failed tests, maybe it's only parallel? Not sure. It's repeatable with this command line:

```
./run_tests --re=fromsub_source -p 4
```

permcody commented:
Actually it looks like you aren't getting output from _any_? Exodiff or CSVDiff test.

Commit:
Fix output issues from the testers

Output was not being returned for all the testers correctly.

Create symbolic link to new RunParallel. This is because we
moved RunParallel but we continue to use 'pathtool' which
places all python modules on a flat namespace, and the old
RunParallel's pyc file will be loaded instead of the new one.

Allow TesterData to return command issued. Multiple calls to
getCommand in RunApp was appending the previous cli_args to
cli_args again each time, thus causing duplicates.

Refs #9280

permcody commented on python/TestHarness/schedulers/TesterData.py:
```diff
@@ -59,6 +61,10 @@ def processCommand(self, command):
 
         return process
 
+    def getCommand(self):
```
I don't really like this for two reasons. First, a command is really part of a tester object, not a the scheduler data object. Second, this getter only works after you've called processCommand, it'll return blank otherwise. 

Caching the command isn't bad, but if you do it that way, it should be in the Tester object. 

The best way would be to make the Tester.getCommand() object not produce any side effects.

milljm commented on python/TestHarness/RunParallel.py:
```diff
@@ -0,0 +1 @@
+schedulers/RunParallel.py
```
Create symbolic link to new RunParallel. This is because we
moved RunParallel but we continue to use 'pathtool' which
places all python modules on a flat namespace, and the old
RunParallel's pyc file will be loaded instead of the new one.

Commit:
Improve long running test reporting

Long running tests are being reported synchronously when we
are only interested in overall TestHarness activity.

Refs #9280

milljm commented on python/TestHarness/schedulers/TesterData.py:
```diff
@@ -111,8 +111,9 @@ def getTiming(self):
+            return self.getEndTime() - self.getStartTime()
+        elif self.getStartTime():
+            # If the test is still running, return current run time instead
+            return max(0.0, clock() - self.getStartTime())
```
No? Cool? who cares?
RUNNING... was always reporting [0.000s] as the time.

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -104,18 +108,6 @@ def skipPrereqs(self):
-               or not tester.getRunnable(self.options):
-                skipped_failed.add(tester)
-        return skipped_failed
-
```
didn't realize I wasn't using this anymore. Cool.

moosebuild commented:
Job [Test debug](https://www.moosebuild.org/job/107300/) on e3f43fc : invalidated by @milljm

something is not working right... taking to long, to much output.

milljm commented:
There is a problem with the TIMEOUT thread timer. Need to re-implement the thread timers.

Commit:
Fix thread timers causing strange timout issues

Restarting the timers was not how we should be handling reporting.

Refs #9280

Commit:
Rework the long running thread timer

Allow the job_container to contain the long running thread timer.

Refs #9280

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/107798/) on e43a9f9 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format a8f782c4f939fdcdff1e5738128a1682a68e8deb`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9335/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9335/style.patch | git apply -v`

Commit:
revert back to minReportTime

Revert back to minReportTime functionality. There are some tests
that use this functionality.
Rename the tester_pool to more accurately describe its function.
Fix a few docstrings with syntax issues.

Refs #9280

moosebuild commented:
Job [Test Intel](https://www.moosebuild.org/job/108125/) on 4c67ccf : invalidated by @milljm

timeouts
